### ISSUE TITLE
Natural language FPs with rule 932236 PL2

### ISSUE THREAD

**Issue by RedXanadu** [coreruleset/coreruleset#3511]
_This issue is part of the ongoing_ quantitative testing _natural language FPs improvement task. See the main issue for details: #3392_

### Description

Rule 932236 struggles with false positives on natural language input. This is because there are many English-language words in the regexp-assemble files for this rule, e.g. "history", "composer", "local", etc.

Rule 932236 is one of the new/overhauled Unix command detection rules in v4.0:
```
Remote Command Execution: Unix Command Injection (command without evasion)
...
#  .932236 (stricter sibling of 932230, 932235, 932250, 932260, PL2,
#       - with and without prefix
#       - words of any length)
```

### Logs

The **biggest offenders** by far were "As " followed by "At ".

Here is a summary of the relevant matched payloads I could find:

```
[data "Matched Data:  Links  found within ARGS_NAMES: Links family-friendly community event from 1-4 p.m. when your library will be transformed into a miniature golf and bowling center for one day only.:  Links family-friendly community event from 1-4 p.m. when your library will be transformed into a miniature golf and bowling center for one day only."]
[data "Matched Data: Arch  found within ARGS:payload: Arch Capital is Watford\\xe2\\x80\\x99s largest shareholder with a 12.6% stake, according to Refinitiv Eikon data."]
[data "Matched Data: As  found within ARGS:payload: As you near your last year of trick or treating, though, sometimes jokes aren\\xe2\\x80\\x99t enough."]
[data "Matched Data: At  found within ARGS:payload: At any moment we can make a decision to change."]
[data "Matched Data: Composer  found within ARGS:payload: Composer Chris Ruenes livens things up, too, with a percussion-heavy score that is playful and offbeat."]
[data "Matched Data: Curl  found within ARGS:payload: Curl up with a good book and snuggly blanket."]
[data "Matched Data: Cut  found within ARGS:payload: Cut orange quarters into crosswise slices about 1/8-inch thick."]
[data "Matched Data: Dash  found within ARGS:payload: Dash (DASH) traded 0.2% higher against the dollar and now trades at $69.27 or 0.00757243 BTC."]
[data "Matched Data: Dialog  found within ARGS:payload: Dialog Semiconductor Plc develops and distributes highly integrated, mixed signal integrated circuits (ICs) for personal, portable, hand-held devices, low energy short-range wireless, LED solid-state lighting, and automotive applications worldwide."]
[data "Matched Data: Expect  found within ARGS:payload: Expect mobile, alternating lane closures Monday through Friday from 7 a.m. to 7 p.m."]
[data "Matched Data: Find  found within ARGS:payload: Find new news sources!\\xe2\\x80\\x9d"]
[data "Matched Data: Get  found within ARGS:payload: Get on course with the rest of the world and set an example of what to do, instead of what not to do."]
[data "Matched Data: Go  found within ARGS:payload: Go early, go hard, go Jacinda."]
[data "Matched Data: Head  found within ARGS:payload: Head of Xbox Phil Spencer was playing the Xbox Series X in 2019, long before it was available to the public."]
[data "Matched Data: History  found within ARGS:payload: History tells us people in the regions, mostly in northern Namibia always enjoyed reading in their local languages."]
[data "Matched Data: Id found within ARGS:payload: Identity politics is the politics of narcissism and ingratitude."]
[data "Matched Data: Id found within ARGS:payload: Idlib province and its surroundings have been a dead end for many people previously displaced by Russian-backed regime operations against former rebel strongholds across the country."]
[data "Matched Data: Last  found within ARGS:payload: Last year, the use of plain Japanese spread to other local governments, including Tokyo, and the Immigration Services Agency decided to create a guideline for using the simpler language form."]
[data "Matched Data: Local  found within ARGS:payload: Local fishing is beneath their concern."]
[data "Matched Data: Look  found within ARGS:payload: Look at you now!"]
[data "Matched Data: Make  found within ARGS:payload: Make contingency plans."]
[data "Matched Data: Mount  found within ARGS:payload: Mount Airy basketball coach Bryan Hayes lectures his team during a third-quarter timeout in the 2018 Frank Spencer Tournament."]
[data "Matched Data: Net  found within ARGS:payload: Net cash outflows last year were $3.968 billion."]
[data "Matched Data: RED  found within ARGS:payload: RED WING \\xe2\\x80\\x94 A former church bookkeeper pleaded guilty Monday to stealing more than $33,000 over three years."]
[data "Matched Data: Self  found within ARGS:payload: Self was second on the team in kills with 11 along with a block and three digs."]
[data "Matched Data: Shadow  found within ARGS:payload: Shadow Treasurer Jim Chalmers has called on the government to release its JobKeeper review."]
[data "Matched Data: Snap  found within ARGS:payload: Snap also said it would stop promoting Mr. Trump\\xe2\\x80\\x99s Snapchat account after determining that his public comments off the site could incite violence."]
[data "Matched Data: Top found within ARGS:payload: Top administration officials, under growing pressure to act, are dismissing the need for immediate, broad-based measures to shore up the economy but have begun suggesting that a smaller-scale response might be necessary."]
[data "Matched Data: Wall  found within ARGS:payload: Wall Street\\xe2\\x80\\x99s benchmark stock index struck an all-time high on Tuesday, having rallied more than 50 per cent from the darkest days of the coronavirus crisis, despite persistent investor unease about the US economy."]
[data "Matched Data: time as  found within ARGS:payload: Abela had attempted to justify the move as necessary at the time as a way of controlling coronavirus infections in the country."]
[data "Matched Data: time as  found within ARGS:payload: We settled into domestic life quite well, both using the downtime as a reset button on our lives."]
[data "Matched Data: time at  found within ARGS:payload: Arshad, 51, said during his time at home, he is working on a few scripts which he intends to direct someday."]
[data "Matched Data: time last  found within ARGS:payload: \\xe2\\x80\\x9cAround this time last year, I called up asking for a bit of information, and they just said they can't release any information,\\xe2\\x80\\x9d he said."]
```

### How to reproduce the misbehavior (-> curl call)

```
curl https://sandbox.coreruleset.org -H 'x-format-output: txt-matched-rules' -H 'x-crs-paranoia-level: 2' --data "payload=Expect mobile, alternating lane closures Monday through Friday from 7 a.m. to 7 p.m."
```

### Your Environment

* CRS version (e.g., v3.3.4): v4.0.0-rc2
* Paranoia level setting (e.g. PL1) : PL 2
* Operating System and version: CRS modsec2-apache container

### Confirmation

[x] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by franbuehler** [comment]
I tested the reported false positives:

- payload=`Expect` mobile -> match
- payload=You `expect` mobile -> no match
- payload=`Wall` street -> match
- payload=The `Wall` street" -> no match

The problem with most of the false positives is the [start of the string `^` evasion prefix](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/regex-assembly/include/unix-shell-evasion-prefix-start-of-string.ra#L7). By removing this `^`,  29 out of the 33 (87%) of the false positives reported above would be gone.
I think, because this start of the string prefix causes so many false positives, we should consider removing it. And I think this is less dangerous than removing the `;`  for example, that is used to concatenate commands.

Then we have a second problem that causes the rest (4 out of 33) of the false positives reported above: the `time` prefix. Unfortunately this prefix is a common English word and in rare cases two combined English words trigger a false positive, like `time at`, `time as` or `time last`.
3 out of the 4 false positives were caused by a `as` or `at`. By adding the two "biggest offenders" `as` and `at`, maybe also `last`, to the exclude file [unix-shell-fps-pl2.ra](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/regex-assembly/exclude/unix-shell-fps-pl2.ra), we could solve most of the false positives. 

I could open a PR solving the problems, but I would like to hear other opinions and if this is still safe or if this causes too many bypasses?

**Comment by dune73** [comment]
https://github.com/coreruleset/coreruleset/blob/v4.0/dev/regex-assembly/include/unix-shell-evasion-prefix-start-of-string.ra#L7 says it's used by 932250 and 932260. Apparently also by more or less unrelated 932236.

It's a lot of false positives, but it's also a PL2 rule.

How useful is the `^` at all? I mean it's relatively easy to add something in front of the command to bypass this, is not it? I know we are catching some of these evasion methods, but other ones go indetected and execute just fine (e.g. `\ifconfig`).

So I think we could remove that but I would prefer to keep the time since this is PL2 and I do not expect many users aiming for a level above.

Just my 2 cents.

**Comment by franbuehler** [comment]
Yes, I've seen this wrong comment and I would fix this too.

Thanks for your opinion about the `^`.

I wouldn't remove the `time` prefix, but I would add `as` and `at` to the exclusion file for pl2.

**Comment by dune73** [comment]
Yes, but still. If we solve 80% of the problem with the `^` prefix, then I think we can leave it at that for 4.0.0 and gain more experience before we rush into the 2nd step.

**Comment by franbuehler** [comment]
Ok, now I see. I thougt I was not clear enough.

**Comment by franbuehler** [comment]
I had a look at this issue again and tested how many of the tests will fail if we remove the [start of the string `^` evasion prefix](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/regex-assembly/include/unix-shell-evasion-prefix-start-of-string.ra#L7). Then the tests 14, 15, 16, 17, 40, 41, 50, 51 are failing. Some of them are BB tests. That's bad I think.

Then I tested what happens if I add the words above to the already existing [exclusion file at pl2](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/regex-assembly/exclude/unix-shell-fps-pl2.ra). And I think this is the way to go. I even think we can add the words starting with a upper case letter -> then no tests will fail. And in natural language the start of a sentence is always an upper case letter. Unix is case sensitive, so we would be safe I think. I'll open a PR, then we can discuss this approach.

**Comment by theseion** [comment]
Honestly, I think we should merge `unix-shell-evasion-prefix.ra` and `unix-shell-evasion-prefix-start-of-string.ra`. The former has received some updates that are missing in the latter, the latter has the additional `^` and `=` matches. I had never analysed the prefixes when I deconstructed them, I had just made sure that we keep them as they were originally written. Looking at them now, I don't see any advantage in having both.

Additionally, the way that `time` is bing used in the `...start-of-string.ra` file is wrong, IMO. It should be identical to `...evasion-prefix.ra` (prefix instead of alterneration).

In conclusion:
- add `=` to `...evasion-prefix.ra`
- remove `...start-of-string.ra`

**Comment by dune73** [comment]
Re: @franbuehler : I think the uppercase thought is great. Never thought about this in relation to FPs!

**Comment by theseion** [comment]
Upper case won't work. File system case sensitivity is the choice of the user. I always use case-insensitive.

**Comment by dune73** [comment]
Hold on, you run a unix system case-insensitive?

I have never heard anyone doing that. Is it a widespread practice?

I'd probably opt to ignore what seems to be an exotic setting and still make use of case sensitivity for regexes (apparently not for PM) when it helps us fight FPs, namely in PL1 and PL2. 

But I am open to hear different opinions on the subject. Maybe I am just not up to date how people run their servers.

**Comment by theseion** [comment]
I've been running macOS case-insensitive for years. Whether it's wide spread, particularly on servers, I don't know. Probably not, but we just don't know.

**Comment by lifeforms** [comment]
The POSIX specification actually allows for case insensitivity, but I think on Linux these filesystems are very uncommon (although ext4 seems to support it) and it is not worth suffering false positives by triggering on it.

**Comment by dune73** [comment]
I've found a bit of information on the subject:

https://www.collabora.com/news-and-blog/blog/2020/08/27/using-the-linux-kernel-case-insensitive-feature-in-ext4/

It seems to be a fairly new addition to ext4 and other filesystems and it is namely used for Windows (Mac if your will) compatibility.

I'd still say we ignore this use case for lower PLs in favor of fighting false positives effectively.

**Comment by theseion** [comment]
As I wrote above, I believe the regular expression is wrong. IMO, the discussion around case sensitivity is moot as long as that's the case.

**Comment by RedXanadu** [comment]
Cross-posting from the PR:

After running the quantitative testing script against this branch (using the same 10k sample set used elsewhere), the changes in this PR take the false positive rate for rule 932236 from:

268 -> 28

**Comment by theseion** [comment]
I've analysed the contents of the prefixes and rule 932236. I still believe that we should merge the prefixes but it's probably better to do that another time. For now we should:
- add `=` to `...evasion-prefix.ra`
- align the "prefix commands" (e.g., `time`) in both files

As for the FPs, I now believe that @franbuehler is right and we should exclude `as`, `at`, etc.. However, we should not add them to `unix-shell-fps-pl2.ra`, as that file is used by multiple rules. Instead, we should add them to a new exclusion file.
I also wouldn't want to remove the `^` match, for the reason @franbuehler has given.

I will update @franbuehler's PR.

**Comment by dune73** [comment]
Thanks for taking this over and looking into it while @franbuehler is on holiday.

**Comment by RedXanadu** [comment]
This is on the cusp of being solved, IMO. Just need to sort out the tests in the PR.
